Issue #2442: adding xdoc check validation

rnveach opened the pull request:
preview of xdoc check validation. no fixes here, will be in separate commit. shouldn't be merged until then.
use reflection alot here.

**Added the following requirements:**
All sections shouldn't end with 'Check'.
All sections should be able to be instantiated.
All sub-sections should follow the order: Description, Properties (optional), Examples, Example of Usage, Package, Parent Module
Following sub-sections can be out of order and optional: Notes, Rule Description, Use Case

Package must be the check's package location.
Parent Module must be TreeWalker if it extends Check, otherwise must be Checker.
Properties is optional, but must exist if there is a bean to be documented, or its tokens should be documented. (default != acceptable != required)

**Still to implement:**
If tokens property is displayed, all tokens must be displayed for type (acceptable) and default value.

@romani 
**Questions:**
1) Should 'Description' sub-section forbid source tags? I feel descriptions should be text only for simplicity. We should have a different section for code examples.
2) Should 'Examples' sub-section show all properties in atleast one example? Exlude tokens?
3) When I build the list of properties, I am going through all sub-classes. Some sub-classes have a property, while the main class sets it to a specific value. Right now these aren't document but could be set for any random reason, so I am making a list of "undocumented" beans for this. This sound ok? You can see my current list in the code.
4) Any other requirements you want me to add?

romani commented:
> Should 'Description' sub-section forbid source tags?

I do not think so, you can write a lot of text, but is always better to show user what code is bad and what code is good.
Lets come to certain problems and make a decision base on them.

> Should 'Examples' sub-section show all properties in atleast one example? Exlude tokens?

I would make few examples:
- default configuration
- some configurations that are likely to be in use.
- configuration with all options to represent power of customization of behaviour.

From user perspective, the more examples the better.

> so I am making a list of "undocumented" beans for this. This sound ok?

Please make an example, I hardly understand what you are talking about.

> Any other requirements you want me to add?

Not for now.
Lets make this issue first, resolve all problems as they become acute.

rnveach commented:
> Please make an example

ReturnCountCheck
http://checkstyle.sourceforge.net/config_coding.html#ReturnCount

It actually has an undocument property, "compileFlags", coming from "AbstractFormatCheck".
The following config is valid, showing that this is indeed a valid property that CS will accept, but there is no valid reason the users should use it or see it documented.
````
	<module name="TreeWalker">
		<module name="ReturnCount">
			<property name="compileFlags" value="2" />
		</module>
	</module>
````

mkordas commented on src/test/java/com/puppycrawl/tools/checkstyle/XDocsPagesTest.java:
```diff
@@ -244,4 +267,230 @@ private static void testCheckstyleXml(String fileName, String code, String unser
+                    // can be in different orders, and completely optional
+                    if (subSectionName.equals("Notes"))
+                        continue;
+                    if (subSectionName.equals("Rule Description"))
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Position literals first in String comparisons](https://www.codacy.com/app/checkstyle/checkstyle/file/1183653445/issues/source?bid=2193633&fileBranchId=2921952#l320)

mkordas commented on src/test/java/com/puppycrawl/tools/checkstyle/XDocsPagesTest.java:
```diff
@@ -244,4 +267,230 @@ private static void testCheckstyleXml(String fileName, String code, String unser
+                    if (subSectionName.equals("Use Case"))
+                        continue;
+
+                    if (subSectionPos == 1 && !subSectionName.equals("Properties")) {
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Position literals first in String comparisons](https://www.codacy.com/app/checkstyle/checkstyle/file/1183653445/issues/source?bid=2193633&fileBranchId=2921952#l326)

mkordas commented on src/test/java/com/puppycrawl/tools/checkstyle/XDocsPagesTest.java:
```diff
@@ -244,4 +267,230 @@ private static void testCheckstyleXml(String fileName, String code, String unser
     }
+
+    @Test
+    public void testAllCheckSections() throws IOException, CheckstyleException,
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Avoid really long methods.](https://www.codacy.com/app/checkstyle/checkstyle/file/1183653445/issues/source?bid=2193633&fileBranchId=2921952#l272)

mkordas commented on src/test/java/com/puppycrawl/tools/checkstyle/XDocsPagesTest.java:
```diff
@@ -244,4 +267,230 @@ private static void testCheckstyleXml(String fileName, String code, String unser
+    }
+
+    private static Object getSubSectionName(int subSectionPos) {
+        switch (subSectionPos) {
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Switch statements should have a default label](https://www.codacy.com/app/checkstyle/checkstyle/file/1183653445/issues/source?bid=2193633&fileBranchId=2921952#l361)

mkordas commented on src/test/java/com/puppycrawl/tools/checkstyle/XDocsPagesTest.java:
```diff
@@ -244,4 +267,230 @@ private static void testCheckstyleXml(String fileName, String code, String unser
+                            .getNodeValue();
+
+                    // can be in different orders, and completely optional
+                    if (subSectionName.equals("Notes"))
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Position literals first in String comparisons](https://www.codacy.com/app/checkstyle/checkstyle/file/1183653445/issues/source?bid=2193633&fileBranchId=2921952#l318)

mkordas commented on src/test/java/com/puppycrawl/tools/checkstyle/XDocsPagesTest.java:
```diff
@@ -244,4 +267,230 @@ private static void testCheckstyleXml(String fileName, String code, String unser
+                    if (subSectionName.equals("Rule Description"))
+                        continue;
+                    // TODO: merge this with description
+                    if (subSectionName.equals("Use Case"))
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Position literals first in String comparisons](https://www.codacy.com/app/checkstyle/checkstyle/file/1183653445/issues/source?bid=2193633&fileBranchId=2921952#l323)

romani commented:
> It actually has an undocument property, "compileFlags", coming from "AbstractFormatCheck".

Good catch. We need to change this Check and extend it from "Check" class only. In your case, lets have hardcoded a list of ignore. Update for Check need to be done in separate issue.

rnveach commented:
**Update.**
Added all requirement tests.
**Fixed all violations.**
All properties that shouldn't be documented accounted for.
Missing XML Fileset for "RegexpSingleline". I'm not sure why this wasn't picked up for the XML validations.
Added override for CustomImportOrderCheck who has missing setter.
Added suppression for how complex XDocsTest is.
AbstractHeaderCheck and SeverityMatchFilter had beans that didn't match up with their setters.

**New Requirements Added:**
Token property should be last.
Token properties have to match default and acceptable token list display. (default/acceptable - required = list).
Token property description must be "tokens to check".
All other properties must have description and type columns filled in. Can't validate default since it can be an empty string.

**Still to do:**
Fix test code complexity and any push violations.
Add new 'TODO's for newly created descriptions and examples. I will see if I can do this.

mkordas commented on src/test/java/com/puppycrawl/tools/checkstyle/XDocsPagesTest.java:
```diff
@@ -244,4 +304,335 @@ private static void testCheckstyleXml(String fileName, String code, String unser
+        return result;
+    }
+
+    private static void validatePropertySection(String fileName, String sectionName,
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Avoid really long methods.](https://www.codacy.com/app/checkstyle/checkstyle/file/1191370016/issues/source?bid=2193633&fileBranchId=2921952#l428)

mkordas commented on src/test/java/com/puppycrawl/tools/checkstyle/XDocsPagesTest.java:
```diff
@@ -244,4 +304,335 @@ private static void testCheckstyleXml(String fileName, String code, String unser
     }
+
+    @Test
+    public void testAllCheckSections() throws Exception {
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Avoid really long methods.](https://www.codacy.com/app/checkstyle/checkstyle/file/1191370016/issues/source?bid=2193633&fileBranchId=2921952#l309)

mkordas commented on src/test/java/com/puppycrawl/tools/checkstyle/XDocsPagesTest.java:
```diff
@@ -244,4 +304,335 @@ private static void testCheckstyleXml(String fileName, String code, String unser
+    private static String getTokenText(int[] tokens, int[] subtraction) {
+        String result;
+
+        if ((Arrays.equals(tokens, TokenUtils.getAllTokenIds())) && (subtraction.length == 0)) {
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Useless parentheses.](https://www.codacy.com/app/checkstyle/checkstyle/file/1191370016/issues/source?bid=2193633&fileBranchId=2921952#l597)

rnveach commented:
@romani 
What is the difference between "Examples" sub-section and "Example of Usage"?
Sometimes we do one or the other.

http://checkstyle.sourceforge.net/config_annotation.html#AnnotationLocation
Sometimes we have xml examples in "Examples" and links to configs in "Example of Usage".

http://checkstyle.sourceforge.net/config_filters.html#SuppressWithNearbyCommentFilter
Other times we have the xml examples in "Example of Usage" and no "Examples".

http://checkstyle.sourceforge.net/config_filters.html#SuppressionFilter
And atleast one with xml examples in both.

I am thinking "Examples" should be a required subsection and "Examples of Usage" is optional, and XML examples should be in "Examples" only.

romani commented:
> What is the difference between "Examples" sub-section and "Example of Usage"?

"Examples of Usage" - links to existing configs with certain number
"Examples" just some examples of code and config and violations.

> Other times we have the xml examples in "Example of Usage" and no "Examples".

Please rename it to "Examples". Now we need certainly differentiate that sections 

> And atleast one with xml examples in both.

Combine in one "Examples" section.

> I am thinking "Examples" should be a required subsection 

yes

> "Examples of Usage" is optional, 

it has to be required too, all Checks and all filters has to be used in our configs over our code, and at least our config should be in "Example of usage".

rnveach commented:
@romani you may start reviewing.
You can view in browser via: http://rnveach.github.io/checkstyle/checks.html
Only one code violation remains from CS. if you have any thoughts on how to fix it, feel free to say.
> XDocsPagesTest.java:342:5: error: Cyclomatic Complexity is 13 (max allowed is 10).

**New Requirements Added:**
"Example of Usage" can't have xml examples.

romani commented:
Please add cyclomatc complexity to suppression and let me merge you PR

rnveach commented:
done.

mkordas commented on src/test/java/com/puppycrawl/tools/checkstyle/XDocsPagesTest.java:
```diff
@@ -244,4 +304,367 @@ private static void testCheckstyleXml(String fileName, String code, String unser
+                + properties, properties.isEmpty());
+    }
+
+    private static void validatePropertySectionProperties(String fileName, String sectionName,
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Avoid long parameter lists.](https://www.codacy.com/app/checkstyle/checkstyle/file/1201264782/issues/source?bid=2193633&fileBranchId=2921952#l482)
